### PR TITLE
Client-side routing

### PR DESCRIPTION
**Pull request by myabc**
https://www.openproject.org/work_packages/7570

### CODE REVIEW

**Comment by tilt** [comment]
should we ignore the CI for a while?

File: app/assets/javascripts/angular/controllers/work-package-details-controller.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  function($scope, $stateParams) {
+
+    $scope.workPackageId = $stateParams.workPackageId
+
+    $scope.$watch('rows', function(rows) {
+      if (rows && rows.length > 0) {
+        var row = $scope.rows.find(function(row) {
+          return row.object.id == $scope.workPackageId;
+        });
+        $scope.workPackage = row ? row.object : {};
```
Comment (tilt) [inline app/assets/javascripts/angular/controllers/work-package-details-controller.js:43]: I'd classically request the work package object from the WorkPackageTableService

File: app/assets/javascripts/angular/controllers/work-packages-controller.js [hunk 2]
```diff
@@ -81,6 +83,30 @@ angular.module('openproject.workPackages.controllers')
...
+        allTypes = projects.map(function(project) {
+          return project.embedded.types;
+        }).reduce(function(a, b) {
+          return a.concat(b);
+        }, []);
+
+        $scope.availableTypes = allTypes; // TODO remove duplicates
+      });
+    }
+  }
```
Comment (tilt) [inline app/assets/javascripts/angular/controllers/work-packages-controller.js:108]: as this is about scope setup it doesn't look wrong to have this code here. But do we could also manage the references in the ProjectService. I'll get an impression of this later

File: app/assets/javascripts/angular/openproject-app.js [hunk 3]
```diff
@@ -27,40 +27,100 @@
...
+  'openproject.messages',
+  'openproject.timeEntries',
+  'ngAnimate',
+  'ngSanitize'
+]);
 
 window.appBasePath = jQuery('meta[name=app_base_path]').attr('content') || '';
 
+// FIXME: replace with actual data
+window.gon = { project_types: [] };
```
Comment (tilt) [inline app/assets/javascripts/angular/openproject-app.js]: doesn't this overwrite any "real" objects passed via gon?

File: app/assets/javascripts/angular/routing.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See doc/COPYRIGHT.rdoc for more details.
+//++
+
+openprojectApp.config(['$stateProvider', '$urlRouterProvider',
```
Comment (tilt) [inline app/assets/javascripts/angular/routing.js]: you broke with the one dependency per line convention you introduced ;)

File: app/assets/javascripts/angular/openproject-app.js (hunk 3 above)
Comment (myabc) [inline app/assets/javascripts/angular/openproject-app.js]: @tilt this can actually be removed now

**Comment by tilt** [comment]
What was that about long Pull Requests?

... Looks good!

Some things aren't clear to me for now, e.g. if we need hash-style URLs, and I haven't seen any ng-view so far, but I'm still a bit lost

File: app/assets/javascripts/angular/controllers/work-package-details-controller.js (hunk 1 above)
Comment (myabc) [inline app/assets/javascripts/angular/controllers/work-package-details-controller.js:43]: @tilt `WorkPackageTableService.getRowsData()` ?

File: app/assets/javascripts/angular/routing.js (hunk 4 above)
Comment (myabc) [inline app/assets/javascripts/angular/routing.js]: @tilt I know! – I can change this, but I don't anticipate introducing new dependencies here.

File: app/assets/javascripts/angular/controllers/work-package-details-controller.js (hunk 1 above)
Comment (tilt) [inline app/assets/javascripts/angular/controllers/work-package-details-controller.js:43]: no, there is no such function yet. Let's just move this logic to the service, but later

**Pull request merged**